### PR TITLE
Do not build and install shared libs when not supported

### LINKED ISSUE #1051: Dune expects .so files generated even when shared libraries are not supported
See ocsigen/lwt#599. @rgrinberg, can you move the most relevant details into a comment on this issue?

### PR DESCRIPTION
**Pull request by ghost**
After this PR, dune no longer tries to build and install shared libraries when this is not supported by the OS. To determine whether this is supported, dune reads the contents of the `Makefile.config` file installed by OCaml and looks for `SUPPORT_SHARED_LIBRARIES`.

This should fix #1051. /cc @aantron, @gruenewa and @lvicentesanchez

### CODE REVIEW

File: src/context.ml [hunk 1]
```diff
@@ -105,13 +106,39 @@ let sexp_of_t t =
...
+    ; "supports_shared_libraries", bool t.supports_shared_libraries
     ; "opam_vars", string_hashtbl string t.opam_var_cache
     ; "ocaml_config", Ocaml_config.sexp_of_t t.ocaml_config
     ; "which", string_hashtbl (option path) t.which_cache
     ]
 
 let compare a b = compare a.name b.name
 
+(* Parse the [`ocamlc -where`/makefile_config] file *)
+module Makefile_config = struct
```
Comment (emillon) [inline src/context.ml]: What do you think about putting this in its own module and making `t` abstract? This is a small addition, but it's part of a large module so it seems worth doing it to keep things manageable.

File: src/lib_rules.ml [hunk 2]
```diff
@@ -340,7 +341,7 @@ module Gen (P : Install_rules.Params) = struct
 
     let dep_graphs = Ocamldep.rules cctx in
 
-    let dynlink = lib.dynlink in
+    let dynlink = lib.dynlink && ctx.supports_shared_libraries in
```
Comment (emillon) [inline src/lib_rules.ml]: This pattern (`lib.dynlink` && `ctx.supports_shared_libraries`) appears 3 times. Is it worth extracting a combinator for it? (I guess `Context` would be the natural place for it)

File: src/context.ml (hunk 1 above)
Comment (diml) [reply src/context.ml]: Looking at this again, this code should clearly go in the ocaml_config library. I moved it there

File: src/lib_rules.ml (hunk 2 above)
Comment (diml) [reply src/lib_rules.ml]: I don't think it would be much clearer than this though

Comment (emillon) [reply src/lib_rules.ml]: Yes, but I'm wondering whether we'll think of adding `&& ctx.supports_shared_libraries` next time we'll use `lib.dynlink`. The current solution works for me, though.

Comment (diml) [reply src/lib_rules.ml]: I guess we could add an API like this:

```ocaml
module Dynlink_supported : sig
  type t
  val of_bool : bool -> t
  module Global : sig
    type t
    val of_bool : bool -> t
  end
  val get : t -> Global.t -> bool
end
```

and change the type of `lib.dynlink` to `Dynlink_supported.t` and the type of `ctx.supports_shared_libraries` to `Dynlink_supported.Global.t`. Then we are sure to catch such errors.

Comment (diml) [reply src/lib_rules.ml]: I did that

File: src/ocaml-config/ocaml_config.ml [hunk 3]
```diff
@@ -209,7 +214,11 @@ module Vars = struct
         | Some i ->
           let x =
             (String.sub line ~pos:0 ~len:i,
-             String.sub line ~pos:(i + 2) ~len:(String.length line - i - 2))
+             let len = String.length line - i - 2 in
+             if len < 0 then
+               ""
+             else
+               String.sub line ~pos:(i + 2) ~len)
```
Comment (rgrinberg) [inline src/ocaml-config/ocaml_config.ml]: Imo, you should just introduce `String.drop`. Otherwise this introduces the possibility for off by 1 errors.

**Review by rgrinberg — APPROVED** [review]
This Dynlink_supported modules seems like its more trouble than its worth. But I don't have a strong opinion on it and it's unlikely to introduce additional maintenance pain

**Comment by diml** [comment]
I guess it serves its purpose. Since we don't test the case where `SUPPORTS_SHARED_LIBRARIES` is `false`, I tend to think we should keep it

**Review by emillon — APPROVED** [review]
I agree that it's a bit verbose but it's easier to refactor that way. Let's give it a go!

**Pull request merged**